File: sdk/sql/query/query.ts
import type { ClientResourceRef } from '@aptre/bldr-sdk/resource/client.js'
import { Resource } from '@aptre/bldr-sdk/resource/resource.js'

import type { SqlValue } from '../../../db/sql/sql.pb.js'
import { SqlQueryResourceServiceClient } from './query_srpc.pb.js'
import type { GetQueryTextResponse, RunQueryResponse } from './query.pb.js'

// SqlQueryTypeID is the world ObjectType id for SQL queries.
export const SqlQueryTypeID = 'sql/query'

// SqlQueryBlockTypeID is the block type id for SQL query roots.
export const SqlQueryBlockTypeID =
  'github.com/s4wave/spacewave/sdk/sql/query.Query'

// SqlQuery is the typed SDK handle for a sql/query object.
export class SqlQuery extends Resource {
  private service: SqlQueryResourceServiceClient

  constructor(resourceRef: ClientResourceRef) {
    super(resourceRef)
    this.service = new SqlQueryResourceServiceClient(resourceRef.client)
  }

  // getQueryText returns query text and target metadata.
  public async getQueryText(
    abortSignal?: AbortSignal,
  ): Promise<GetQueryTextResponse> {
    return this.service.GetQueryText({}, abortSignal)
  }

  // setQueryText updates query text and target metadata.
  public async setQueryText(
    sqlText: string,
    dialectHint: string,
    targetDbObjectKey: string,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.SetQueryText(
      { sqlText, dialectHint, targetDbObjectKey },
      abortSignal,
    )
  }

  // setParameters updates positional bind arguments.
  public async setParameters(
    parameters: SqlValue[],
    abortSignal?: AbortSignal,
  ): Promise<void> {
    await this.service.SetParameters({ parameters }, abortSignal)
  }

  // run executes the query and creates a query result object.
  public async run(
    maxRows = 0,
    abortSignal?: AbortSignal,
  ): Promise<RunQueryResponse> {
    return this.service.Run({ maxRows }, abortSignal)
  }
}

File: sdk/sql/query/query.go
package s4wave_sql_query

import (
	"context"

	"github.com/s4wave/spacewave/db/block"
	"github.com/s4wave/spacewave/db/blocktype"
	"github.com/s4wave/spacewave/db/bucket"
	"github.com/s4wave/spacewave/db/world"
	s4wave_sql "github.com/s4wave/spacewave/sdk/sql"
)

// SqlQueryTypeID is the world ObjectType id for SQL queries.
const SqlQueryTypeID = "sql/query"

// SqlQueryBlockTypeID is the block type id for SQL query roots.
const SqlQueryBlockTypeID = "github.com/s4wave/spacewave/sdk/sql/query.Query"

// SqlQueryBlockType constructs SQL query root blocks for typed cursor writes.
var SqlQueryBlockType = blocktype.NewBlockType(
	SqlQueryBlockTypeID,
	func() *Query { return &Query{} },
)

// NewQueryBlock constructs a SQL query block.
func NewQueryBlock() block.Block {
	return &Query{}
}

// ReadQueryRoot reads a SQL query object's root.
func ReadQueryRoot(ctx context.Context, ws world.WorldState, objectKey string) (*Query, error) {
	obj, err := world.MustGetObject(ctx, ws, objectKey)
	if err != nil {
		return nil, err
	}
	return ReadQueryObjectRoot(ctx, obj)
}

// ReadQueryObjectRoot reads a SQL query root from an object state.
func ReadQueryObjectRoot(ctx context.Context, obj world.ObjectState) (*Query, error) {
	var query *Query
	_, _, err := world.AccessObjectState(ctx, obj, false, func(bcs *block.Cursor) error {
		var err error
		query, err = block.UnmarshalBlock[*Query](ctx, bcs, NewQueryBlock)
		if err != nil {
			return err
		}
		if query == nil {
			return world.ErrObjectNotFound
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return query, nil
}

// WriteQueryRootRef writes a SQL query root block and returns its ref.
func WriteQueryRootRef(ctx context.Context, ws world.WorldState, query *Query) (*bucket.ObjectRef, error) {
	return world.AccessObject(ctx, ws.AccessWorldState, nil, func(bcs *block.Cursor) error {
		bcs.SetBlock(query, true)
		return nil
	})
}

// SyncTargetDbQuad replaces the query's target database graph link.
func SyncTargetDbQuad(ctx context.Context, ws world.WorldState, objectKey string) error {
	query, err := ReadQueryRoot(ctx, ws, objectKey)
	if err != nil {
		return err
	}
	quads, err := ws.LookupGraphQuads(
		ctx,
		world.NewGraphQuadWithKeys(objectKey, s4wave_sql.PredSqlQueryAgainst.String(), "", ""),
		0,
	)
	if err != nil {
		return err
	}
	for _, q := range quads {
		if err := ws.DeleteGraphQuad(ctx, q); err != nil {
			return err
		}
	}
	if query.GetTargetDbObjectKey() == "" {
		return nil
	}
	return ws.SetGraphQuad(ctx, world.NewGraphQuadWithKeys(
		objectKey,
		s4wave_sql.PredSqlQueryAgainst.String(),
		query.GetTargetDbObjectKey(),
		"",
	))
}

// MarshalBlock marshals the SQL query root.
func (q *Query) MarshalBlock() ([]byte, error) {
	return q.MarshalVT()
}

// UnmarshalBlock unmarshals the SQL query root.
func (q *Query) UnmarshalBlock(data []byte) error {
	return q.UnmarshalVT(data)
}

// Validate performs cursory checks on the SQL query root.
func (q *Query) Validate() error {
	return nil
}

// _ is a type assertion.
var _ block.Block = (*Query)(nil)
